rir/src/compiler/opt/inline_force_prom.cpp

```diff
@@ -0,0 +1,78 @@
+#include "../analysis/available_checkpoints.h"
+#include "../analysis/force_dominance.h"
+#include "../parameter.h"
+#include "../pir/pir_impl.h"
+#include "compiler/util/bb_transform.h"
+#include "compiler/util/safe_builtins_list.h"
+#include "pass_definitions.h"
+#include "utils/Map.h"
+#include "utils/Set.h"
+#include <functional>
+#include <unordered_map>
+#include <unordered_set>
+
+/*
+For each call, if we can statically know the function at a call and the function
+is annotated as depromise, we add a Force instruction for each argument right
+before the call, and we pass the forced values instead. By adding a force, we
+are sure that the corresponding MkArgs are in scope. It means that promises will
+be inlined by a future pass (promise inliner). Corresponding MkArg instructions
+will be later on removed by dead code elimination.
+
+*/
+
+namespace rir {
+namespace pir {
+
+bool InlineForcePromises::apply(Compiler&, ClosureVersion* cls, Code* code,
+                                LogStream& log) const {
+
+    bool anyChange = false;
+
+    Visitor::run(code->entry, [&](BB* bb) {
+        auto ip = bb->begin();
+        while (ip != bb->end()) {
+            auto next = ip + 1;
+
+            if (auto call = CallInstruction::CastCall(*ip)) {
+
+                auto cls = call->tryGetCls();
+
+                if (cls) {
+
+                    auto functionVersion = cls->rirFunction();
+                    if (functionVersion->flags.contains(
+                            rir::Function::Flag::Annotated)) {
```

Comment:
@o-  here is the use of the flag for optimizations. Is this what you were referring to in your comment?

Reply:
no. I mean that the type of function arguments should be adjusted for functions with the depromised annotation